src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,406 @@
+            cf_contents,
+        }
+    }
+
+    pub fn get_cf(&self, cf: CfName) -> Arc<RwLockTree> {
+        let index = self
+            .cf_names
+            .iter()
+            .position(|&c| c == cf)
+            .expect("Not exist CF!");
+        self.cf_contents[index].clone()
+    }
+}
+
+impl Default for BTreeEngine {
+    fn default() -> Self {
+        let cfs = &[CF_WRITE, CF_DEFAULT, CF_LOCK];
+        Self::new(cfs)
+    }
+}
+
+impl Engine for BTreeEngine {
+    type Iter = BTreeEngineIterator;
+    type Snap = BTreeEngineSnapshot;
+
+    fn async_write(
+        &self,
+        _ctx: &Context,
+        modifies: Vec<Modify>,
+        cb: EngineCallback<()>,
+    ) -> EngineResult<()> {
+        if modifies.is_empty() {
+            return Err(EngineError::EmptyRequest);
+        }
+        cb((CbContext::new(), write_modifies(&self, modifies)));
+
+        Ok(())
+    }
+    /// warning: It returns a fake snapshot whose content will be affected by the later modifies!
+    fn async_snapshot(&self, _ctx: &Context, cb: EngineCallback<Self::Snap>) -> EngineResult<()> {
+        cb((CbContext::new(), Ok(BTreeEngineSnapshot::new(&self))));
+        Ok(())
+    }
+}
+
+impl Debug for BTreeEngine {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "BTreeEngine []",)
+    }
+}
+
+pub struct BTreeEngineIterator {
+    tree: Arc<RwLockTree>,
+    cur_key: Option<Key>,
+    cur_value: Option<Value>,
+    valid: bool,
+    bounds: (Bound<Key>, Bound<Key>),
+}
+
+impl BTreeEngineIterator {
+    pub fn new(tree: Arc<RwLockTree>, iter_opt: IterOption) -> BTreeEngineIterator {
+        let lower_bound = match iter_opt.lower_bound() {
+            None => Unbounded,
+            Some(key) => Included(Key::from_raw(key)),
+        };
+
+        let upper_bound = match iter_opt.upper_bound() {
+            None => Unbounded,
+            Some(key) => Excluded(Key::from_raw(key)),
+        };
+        let bounds = (lower_bound.clone(), upper_bound.clone());
+        Self {
+            tree: tree.clone(),
+            cur_key: None,
+            cur_value: None,
+            valid: false,
+            bounds,
+        }
+    }
+
+    fn seek_and_valid(&mut self, range: (Bound<Key>, Bound<Key>), next_back: bool) -> bool {
+        let tree = self.tree.read().unwrap();
+        let mut range = tree.range(range);
+        let item = if next_back {
+            range.next_back()
+        } else {
+            range.next()
+        };
+        match item {
+            Some((k, v)) if self.bounds.contains(k) => {
+                self.cur_key = Some(k.clone());
+                self.cur_value = Some(v.clone());
+                self.valid = true;
+            }
+            _ => {
+                self.valid = false;
+            }
+        }
+        self.valid()
+    }
+}
+
+impl Iterator for BTreeEngineIterator {
+    fn next(&mut self) -> bool {
+        let range = (Excluded(self.cur_key.clone().unwrap()), Unbounded);
```

Comment:
Can this extra `clone` be avoided? These extra cost will affect the effectiveness of layered benchmarks so we should avoid them as more as possible.

Reply:
No, I had tried to reduce the `clone()` in the whole file, but failed. `range` will get the ownership of `self.cur_key` which will be used in next, so I have to clone it.😂

Reply:
What if `range` takes a reference instead? According to [BTreeMap::range's example](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) it actually only needs a Range over reference.

This may be hard to implement because it will introduce life time to your Iterator. However you can have a try.

Reply:
I tried the whole of last week but failed. It's easy to hold a `range` reference from a raw `BTreeMap`, but when it comes to `Arc<RwLock<BTreeMap>>` there seems no solution.😅 `range` refers from `BTreeEngine` through `ReadGuard` which is a temporary, and it will cause a dangling reference. If I keep `ReadGuard` to the current struct, then the current struct will lose `Send` trait, because of the `!Send` trait of ReadGuard [here](https://doc.rust-lang.org/src/std/sync/rwlock.rs.html#104), and then `BTreeEngine` will be thread unsafe. Any idea?